source/speech/__init__.py

```diff
@@ -1910,6 +1910,18 @@ def getControlFieldSpeech(  # noqa: C901
 		out = []
 		if ariaCurrent:
 			out.extend(ariaCurrentSequence)
+		# Speak expanded / collapsed / level for treeview items (in ARIA treegrids)
+		if role == controlTypes.ROLE_TREEVIEWITEM:
+			if controlTypes.STATE_EXPANDED in states:
+				out.extend(
+					getPropertiesSpeech(reason=reason, states={controlTypes.STATE_EXPANDED}, _role=role)
+				)
+			elif controlTypes.STATE_COLLAPSED in states:
+				out.extend(
+					getPropertiesSpeech(reason=reason, states={controlTypes.STATE_COLLAPSED}, _role=role)
+				)
+			if levelSequence:
+				out.extend(levelSequence)

```

Comment:
Given the complexity (and suppressed warning) for `getControlFieldSpeech` can this be simplified?

Reply:
I can't think of a way currently... and that code I added looks pretty simple to me... unless I'm missing your point?

Reply:
We use mccabe to check for code complexity. It is a quantitative measure of the number of linearly independent paths through the source code. The code complexity check essentially adds points for branching and loops.
You can read more about it here: https://en.wikipedia.org/wiki/Cyclomatic_complexity

I think to reduce the complexity of this system, we'll need to work towards reducing the responsibilities of functions like `getPropertiesSpeech`, splitting them up into more specific pieces.

As an interim, working towards this goal, this block might be replaced with something like:
 ```suggestion
		# Speak expanded / collapsed / level for treeview items (in ARIA treegrids)
		if role == controlTypes.ROLE_TREEVIEWITEM:
			out.extend(
				getExpandedCollapsedLevelForTreeViewItems(reason, states, role)
			)
```

Reply:
And what about level? levelSequence would also need to be passed to that function. 
Personally, I think that that change is actually harder to read and reason about as it does not fit in with the structure of the rest of the outer function. 
I do totally agree however that getcontrolFieldSpeech does need to be completely redesigned at some point, and perhaps split up... but I don't really feel it is right to do it little bit by little bit. I find that more confusing than less sorry.
I think we may even end up finding that a function isn't the right data structure for all of this anyway... it may in fact be something more declarative.

Reply:
>And what about level?

Yep, I forgot to add it.

> getcontrolFieldSpeech really does need to be completely redesigned at some point,

The problem with this is that it's always too much work to ever get started.

> I personally find that more confusing than less, but that might be just me.

It might be in some ways, like having to navigate to the function to know what it does. But in other ways you know it's state and logic is limited to it's return value and params. Knowing this makes refactoring easier, it becomes easier to untangle the dependencies and separate parts of code that aren't really dependent on each other.